### PR TITLE
Enable LLDB tests in Linux pre-merge CI

### PR DESCRIPTION
**Pull request by Endilll**
This patch removes LLDB from a list of projects that are excluded from building and testing on pre-merge CI on Linux.

Windows environment needs to be prepared in order to test LLDB (https://github.com/llvm/llvm-project/pull/94208#issuecomment-2146256857), but we don't have enough maintenance resources to do that at the moment.

Because LLDB has been in the list of projects that need to be tested on Clang changes, this PR make this happen on Linux. This seems to be the consensus in the discussion of this PR.

### CODE REVIEW

**Comment by Michael137** [comment]
Worth a shot i guess. Though it could be somewhat disruptive if we do in fact have flaky tests. I don't think we need to run *all* tests. Should we just run the tests for the LLDB C++ language plugin (i.e., the part of LLDB that invokes Clang and the JIT). And then we could expand the set of tests from there?

**Comment by Endilll** [comment]
> Should we just run the tests for the LLDB C++ language plugin (i.e., the part of LLDB that invokes Clang and the JIT). And then we could expand the set of tests from there later?

In the bash script that currently handles this, "targets that we need to check if lldb is tested" is expressible, but "targets that we need to check if lldb is tested on clang changes" is not, and I consider it already too complicated for a bash script.

**Comment by Endilll** [comment]
First Windows CI results: LLDB unit tests do not even build.
```
FAILED: tools/lldb/unittests/Process/gdb-remote/CMakeFiles/ProcessGdbRemoteTests.dir/GDBRemoteCommunicationClientTest.cpp.obj 
sccache C:\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe  /nologo /TP -DBUILD_EXAMPLES -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_ENABLE_EXTENDED_ALIGNED_STORAGE -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\lldb\unittests\Process\gdb-remote -IC:\ws\src\lldb\unittests\Process\gdb-remote -IC:\ws\src\lldb\include -Itools\lldb\include -Iinclude -IC:\ws\src\llvm\include -IC:\ws\src\llvm\..\clang\include -Itools\lldb\..\clang\include -IC:\ws\src\lldb\source -IC:\ws\src\lldb\unittests -IC:\ws\src\third-party\unittest\googletest\include -IC:\ws\src\third-party\unittest\googlemock\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -MD   -wd4018 -wd4068 -wd4150 -wd4201 -wd4251 -wd4521 -wd4530 -wd4589  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Fotools\lldb\unittests\Process\gdb-remote\CMakeFiles\ProcessGdbRemoteTests.dir\GDBRemoteCommunicationClientTest.cpp.obj /Fdtools\lldb\unittests\Process\gdb-remote\CMakeFiles\ProcessGdbRemoteTests.dir\ /FS -c C:\ws\src\lldb\unittests\Process\gdb-remote\GDBRemoteCommunicationClientTest.cpp
C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\future(196): error C2512: 'llvm::ErrorOr<llvm::MD5::MD5Result>': no appropriate default constructor available
C:\ws\src\llvm\include\llvm/Support/FileSystem.h(432): note: see declaration of 'llvm::ErrorOr<llvm::MD5::MD5Result>'
C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\future(193): note: while compiling class template member function 'std::_Associated_state<_Ty>::_Associated_state(std::_Deleter_base<_Ty> *)'
        with
        [
            _Ty=llvm::ErrorOr<llvm::MD5::MD5Result>
        ]
C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\future(464): note: see reference to function template instantiation 'std::_Associated_state<_Ty>::_Associated_state(std::_Deleter_base<_Ty> *)' being compiled
        with
        [
            _Ty=llvm::ErrorOr<llvm::MD5::MD5Result>
        ]
C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\future(774): note: see reference to class template instantiation 'std::_Associated_state<_Ty>' being compiled
        with
        [
            _Ty=llvm::ErrorOr<llvm::MD5::MD5Result>
        ]
C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\future(769): note: while compiling class template member function '_Ty &std::_State_manager<_Ty>::_Get_value(void) const'
        with
        [
            _Ty=llvm::ErrorOr<llvm::MD5::MD5Result>
        ]
C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\future(884): note: see reference to function template instantiation '_Ty &std::_State_manager<_Ty>::_Get_value(void) const' being compiled
        with
        [
            _Ty=llvm::ErrorOr<llvm::MD5::MD5Result>
        ]
C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\future(860): note: see reference to class template instantiation 'std::_State_manager<_Ty>' being compiled
        with
        [
            _Ty=llvm::ErrorOr<llvm::MD5::MD5Result>
        ]
C:\ws\src\lldb\unittests\Process\gdb-remote\GDBRemoteCommunicationClientTest.cpp(598): note: see reference to class template instantiation 'std::future<llvm::ErrorOr<llvm::MD5::MD5Result>>' being compiled
```

**Comment by Michael137** [comment]
> > Should we just run the tests for the LLDB C++ language plugin (i.e., the part of LLDB that invokes Clang and the JIT). And then we could expand the set of tests from there later?
> 
> In the bash script that currently handles this, "targets that we need to check if lldb is tested" is expressible, but "targets that we need to check if lldb is tested on clang changes" is not, and I consider it already too complicated for a bash script.

Ah i see, could you point me to that part of the bash script?

**Comment by Endilll** [comment]
@Michael137 This is for Linux. Windows works the same way just below those lines: https://github.com/llvm/llvm-project/blob/9a7bd8a60f03595be5d42315790df6d409f81091/.ci/generate-buildkite-pipeline-premerge#L253-L255

**Comment by Michael137** [comment]
> @Michael137 This is for Linux. Windows works the same way just below those lines:
> 
> https://github.com/llvm/llvm-project/blob/9a7bd8a60f03595be5d42315790df6d409f81091/.ci/generate-buildkite-pipeline-premerge#L253-L255

Ah, so what would the issue be with just adding more specific `check-lldb-*` targets, rather than all of `check-lldb`?

**Comment by Endilll** [comment]
> > @Michael137 This is for Linux. Windows works the same way just below those lines:
> > https://github.com/llvm/llvm-project/blob/9a7bd8a60f03595be5d42315790df6d409f81091/.ci/generate-buildkite-pipeline-premerge#L253-L255
> 
> Ah, so what would the issue be with just adding more specific `check-lldb-*` targets, rather than all of `check-lldb`?

Those more specific targets will be used instead of `check-lldb` every time lldb is tested, e.g. when there are just lldb changes in the PR, and Clang is not touched.

**Comment by Endilll** [comment]
Linux results are ready. `SymbolFile/DWARF/x86/module-ownership.mm` is crashing.
While unresolved tests are not considered failed, if you grep for `FAILED`, they add a lot of noise that would make harder for people to realize what the actual reason for CI failure.
All in all, I consider LLDB tests not to be in a good enough shape to add them to our CI in their entirety. Enabling known good subsets is an option, though.
https://buildkite.com/llvm-project/github-pull-requests/builds/69210#018fddab-f349-4495-b5fa-7fa2c329a73f
```
Unresolved Tests (21):
  lldb-api :: commands/expression/multiline-completion/TestMultilineCompletion.py
  lldb-api :: commands/expression/multiline-navigation/TestMultilineNavigation.py
  lldb-api :: commands/gui/basic/TestGuiBasic.py
  lldb-api :: commands/gui/breakpoints/TestGuiBreakpoints.py
  lldb-api :: commands/gui/spawn-threads/TestGuiSpawnThreads.py
  lldb-api :: commands/gui/viewlarge/TestGuiViewLarge.py
  lldb-api :: driver/batch_mode/TestBatchMode.py
  lldb-api :: driver/job_control/TestJobControl.py
  lldb-api :: driver/quit_speed/TestQuitWithProcess.py
  lldb-api :: functionalities/breakpoint/breakpoint_callback_command_source/TestBreakpointCallbackCommandSource.py
  lldb-api :: functionalities/progress_reporting/TestTrimmedProgressReporting.py
  lldb-api :: iohandler/autosuggestion/TestAutosuggestion.py
  lldb-api :: iohandler/completion/TestIOHandlerCompletion.py
  lldb-api :: iohandler/resize/TestIOHandlerResize.py
  lldb-api :: iohandler/sigint/TestIOHandlerPythonREPLSigint.py
  lldb-api :: iohandler/sigint/TestProcessIOHandlerInterrupt.py
  lldb-api :: iohandler/stdio/TestIOHandlerProcessSTDIO.py
  lldb-api :: iohandler/unicode/TestUnicode.py
  lldb-api :: repl/clang/TestClangREPL.py
  lldb-api :: terminal/TestEditline.py
  lldb-api :: terminal/TestSTTYBeforeAndAfter.py

********************
Failed Tests (1):
  lldb-shell :: SymbolFile/DWARF/x86/module-ownership.mm
```

**Comment by Endilll** [comment]
@DavidSpickett I made the changes you suggested. Let's see how this Linux run fares.
Any thoughts on https://github.com/llvm/llvm-project/pull/94208#issuecomment-2144949637 ?

**Comment by DavidSpickett** [comment]
> First Windows CI results: LLDB unit tests do not even build.

Linaro's Windows bot uses clang-cl, but I do remember something like this there too. I will do a local build with msvc and see what I can do.

**Comment by DavidSpickett** [comment]
> I'll do that, but I think LLDB should follow the example of MLIR here

https://github.com/llvm/llvm-project/pull/94220

**Comment by Endilll** [comment]
Linux is passing now in 30 minutes. It's not clear how much overhead LLDB actually added, because other Linux builds on Clang changes are anywhere between 22 and 28 minutes.
I'm waiting for https://github.com/llvm/llvm-project/pull/94220 and news regarding failed MSVC build.

**Comment by JDevlieghere** [comment]
This is awesome, thank you for working on this!

**Comment by Endilll** [comment]
CC @omjavaid https://github.com/llvm/llvm-project/pull/94208#issuecomment-2144949637

**Comment by asl** [comment]
Tagging @slydiman @vvereschaka

**Comment by slydiman** [comment]
> First Windows CI results: LLDB unit tests do not even build.

Please note the following for Windows setup:

- We used cmake 3.25.3 because we got some issues with the latest cmake.
- python prerequisites are: `python -m pip install swig pexpect psutil`.
- Most gnu tools may be used from `git` distributive, PATH env must contain C:\Program Files\Git\usr\bin
- lldb API tests require `make` which can be installed via Chocolatey `choco install make`.
- lldb for Windows host must be compiled with `CMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDLL` (or MultiThreadedDebugDLL in case of CMAKE_BUILD_TYPE=Debug), otherwise python and lldb will use different UCRT instances and python will crash with the exit code 0xC0000409 (STATUS_STACK_BUFFER_OVERRUN) because of an assert in UCRT.

**Comment by Endilll** [comment]
> > First Windows CI results: LLDB unit tests do not even build.
> 
> Please note the following for Windows setup:
> 
>     * We used cmake 3.25.3 because we got some issues with the latest cmake.
> 
>     * python prerequisites are: `python -m pip install swig pexpect psutil`.
> 
>     * Most gnu tools may be used from `git` distributive, PATH env must contain C:\Program Files\Git\usr\bin
> 
>     * lldb API tests require `make` which can be installed via Chocolatey `choco install make`.
> 
>     * lldb for Windows host must be compiled with `CMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDLL` (or MultiThreadedDebugDLL in case of CMAKE_BUILD_TYPE=Debug), otherwise python and lldb will use different UCRT instances and python will crash with the exit code 0xC0000409 (STATUS_STACK_BUFFER_OVERRUN) because of an assert in UCRT.

Thank you for the information. This is a tall order given that Windows pre-submit CI maintenance is currently on the shoulders of a single person. We should reconsider once we move to GitHub actions, and more people can do maintenance and customizations like those, but for now I'm moving ahead without Windows.

**Comment by DavidSpickett** [comment]
> MSVC 14.29.30133 is VS 2019. We are using VS 2022 with the latest Windows SDK to build lldb w/o problems.

I also couldn't reproduce it locally with `19.34.31937`, but I may still be able to reproduce by making my own version of this PR so I can use the CI's version.

**Comment by DavidSpickett** [comment]
https://github.com/llvm/llvm-project/pull/94325 will get us past the compiler error.

Some things I noticed testing that:
* There is an lld installed on the machine but it's in a path with spaces so lit can't run it. Generally we'd build lld along with lldb and use that in the test suite, when I did that it worked.
* LLDB is not being built with scripting enabled, so a lot of the API tests won't be run.

For a completed run I got these results:
https://buildkite.com/llvm-project/github-pull-requests/builds/69540#018fe2bf-1cf8-49f7-a680-f44bf01dbaec
```
Failed Tests (4):
  lldb-shell :: Process/Windows/process_load.cpp
  lldb-shell :: SymbolFile/DWARF/x86/dwp-separate-debug-file.cpp
  lldb-shell :: Target/dependent-modules-nodupe-windows.test
  lldb-shell :: Unwind/windows-unaligned-x86_64.test

Testing Time: 38.08s
Total Discovered Tests: 2901
  Unsupported      : 1337 (46.09%)
  Passed           : 1543 (53.19%)
  Expectedly Failed:   17 (0.59%)
  Failed           :    4 (0.14%)
```

`dwp-separate-debug-file.cpp` I have fixed, it needed to require Python. `windows-unaligned-x86_64.test` just fails to breakpoint, I don't know why. The rest are something to do with `kernel32.dll` either not being able to load it, or it not showing up in the image list for a debugged program.

I have never seen that before but it does sound like something a security setting might effect.

**Review by AaronBallman — COMMENTED** [review]
I think it makes a lot of sense for lldb to have a precommit CI pipeline which tests changes to lldb. I don't think we're at a point where it would make sense to enable lldb precommit CI testing for changes to clang, though.

For starters, the false positive rate from flaky lldb test failures is pretty high:

https://lab.llvm.org/buildbot/#/builders/17/builds/53939
https://lab.llvm.org/buildbot/#/builders/68/builds/75367
https://lab.llvm.org/buildbot/#/builders/219/builds/11908
https://lab.llvm.org/buildbot/#/builders/219/builds/11819
https://lab.llvm.org/buildbot/#/builders/219/builds/11810
https://lab.llvm.org/buildbot/#/builders/96/builds/58350

But also, changes in Clang that break lldb do not necessarily mean that a PR is not ready to land. There are conforming changes that can be made to Clang which lldb needs to react to as a downstream consumer of Clang as a library rather than a PR author needing to react to it. (e.g., adding a new AST node to Clang or changing diagnostic wording/behavior). We should try to come up with and document a policy about what the expectations are regarding Clang and lldb interactions before we add lldb to Clang's precommit CI pipeline.

**Comment by mizvekov** [comment]
> I think it makes a lot of sense for lldb to have a precommit CI pipeline which tests changes to lldb. I don't think we're at a point where it would make sense to enable lldb precommit CI testing for changes to clang, though.
> 
> For starters, the false positive rate from flaky lldb test failures is pretty high:

I think we only need a subset of lldb tests for pure clang changes, and my experience is that those aren't as flaky.

> But also, changes in Clang that break lldb do not necessarily mean that a PR is not ready to land. There are conforming changes that can be made to Clang which lldb needs to react to as a downstream consumer of Clang as a library rather than a PR author needing to react to it. (e.g., adding a new AST node to Clang or changing diagnostic wording/behavior). We should try to come up with and document a policy about what the expectations are regarding Clang and lldb interactions before we add lldb to Clang's precommit CI pipeline.

Yeah clarifying this would be helpful. In my past experience, this is not the expectation lldb maintainers have. I have had clang patches reverted only because they broke lldb tests on pure diagnostics / type printing improvements.

**Comment by adrian-prantl** [comment]
> But also, changes in Clang that break lldb do not necessarily mean that a PR is not ready to land. There are conforming changes that can be made to Clang which lldb needs to react to as a downstream consumer of Clang as a library rather than a PR author needing to react to it. (e.g., adding a new AST node to Clang or changing diagnostic wording/behavior). We should try to come up with and document a policy about what the expectations are regarding Clang and lldb interactions before we add lldb to Clang's precommit CI pipeline.

I would not agree with this _specific_ wording. We have in the past and will in the future revert commits that break existing tests elsewhere in the LLVM project. If a patch to Clang breaks an existing test in LLDB, it's up to the author of the patch to work with the LLDB project members to find path forward so the patch can be merged without introducing a breakage on the bots. The specifics of that will depend on the situation (it could be supporting the feature in LLDB, changing the patch to Clang, updating an overconstrained test, XFAILing a test if there's a really good reason to do so, ...), but it's generally not okay to land a patch that breaks existing tests in the LLVM project.

**Comment by JDevlieghere** [comment]
+1 to what Adrian said. Taht certainly does **not** match my expectations. I'm not sure why there would be a different policy for LLDB than for any of the other monorepo projects.

**Comment by AaronBallman** [comment]
> I would not agree with this specific wording. We have in the past and will in the future revert commits that break existing tests elsewhere in the LLVM project. If a patch to Clang breaks an existing test in LLDB, it's up to the author of the patch to work with the LLDB project members to find path forward so the patch can be merged without introducing a breakage on the bots. The specifics of that will depend on the situation (it could be supporting the feature in LLDB, changing the patch to Clang, updating an overconstrained test, XFAILing a test if there's a really good reason to do so, ...), but it's generally not okay to land a patch that breaks existing tests in the LLVM project.

It's good that we have this discussion then, thank you! FWIW, this scenario is the kind of one I am talking about:

> I have had clang patches reverted only because they broke lldb tests on pure diagnostics / type printing improvements.

I have also had this exact experience and this is the scenario we need to avoid. We (lldb + clang folks) had discussed this previously (I can try to dig up the link if that's helpful) and the eventual decision we came to was that conforming changes to Clang which break lldb do not always qualify for an automatic revert because a bot went red. e.g., Clang can change the wording of a diagnostic and a monorepo "downstream" (like lldb or libc++) may need to react to address the issue without reverting the Clang change. That said, there's definitely an expectation that the Clang developer will let the lldb (or libc++, etc) folks know about the changes and try to avoid disruption as much as possible.

To hopefully help focus the need (not exhaustive lists):

Revert is appropriate when:
* lldb testing exposes a crash or assert in Clang
* the change in Clang is semantically incorrect

A revert is not appropriate when:
* lldb tests fail because AST printing has changed in Clang and the printed output is valid
* lldb tests fail because the wording of a diagnostic has changed in Clang

Examples of gray area are:
* Clang changes the layout of an object and lldb tests start failing
* Clang changes and an lldb test fails but it's not clear why the test is now failing

The key takeaway is: lldb bots going red because of changes in Clang *does not* qualify for an automatic revert in all circumstances, but definitely does qualify in others. To feel comfortable adding lldb to Clang's precommit CI, we need to ensure that it's clear to contributors (including folks new to the community) which failures qualify as "your patch cannot be merged" and which failures qualify as "please let lldb folks know they need to react to these changes".

**Comment by adrian-prantl** [comment]
> A revert is not appropriate when:
> lldb tests fail because AST printing has changed in Clang and the printed output is valid

Following the logic in this argument, would you also say it's appropriate for someone to commit an LLVM patch that changes how textual LLVM IR is printed in a way that it breaks existing tests in Clang — without updating those tests?

As far as I am concerned I would expect someone making such a change to at the very least update the Clang tests to the new format, but even better to have a discussion with the community whether the benefits of this change warrant the churn in the tests.

**Comment by AaronBallman** [comment]
> Following the logic in this argument, would you also say it's appropriate for someone to commit an LLVM patch that changes how textual LLVM IR is printed in a way that it breaks existing tests in Clang — without updating those tests?

Possibly! I'm talking about relying on implementation details of a project and I think the textual representation of LLVM IR isn't necessarily an implementation detail of LLVM. But if the change was to an implementation detail and LLVM had a need to make the change, it's on the Clang community to react to that change.

> As far as I am concerned I would expect someone making such a change to at the very least update the Clang tests to the new format, but even better to have a discussion with the community whether the benefits of this change warrant the churn in the tests.

I'm not trying to suggest that Clang can change anything it wants and lldb needs to just deal with it. I'm saying that Clang needs the ability to change our internal implementation details without *requiring* a PR author to update downstream projects relying on those implementation details. We should work together to reduce the amount of burden between both projects of course, but if a downstream project is expecting a very specific wording for a diagnostic message, Clang developers should not have their (correct) changes reverted because they updated the diagnostic wording and it broke a downstream test.

Would it be useful for us to have a conference call to discuss this in more detail to make sure we're all clear and comfortable?

**Comment by AaronBallman** [comment]
I had a nice chat with @ldionne and @nikic today and we touched on the topic of precommit CI for lldb being triggered from Clang, and that resulted in a change of opinion and perhaps a better way for me to phrase what I'm after and how it relates to this PR.

I think that what we ultimately need is improved coordination between projects when a change has wider-reaching impacts. If a change to an implementation detail in Clang breaks an lldb test, we don't want to land the change and disrupt the community. But the developer making the change may not be able to fix the issues outside of Clang (lack of expertise, time constraints, etc) and we don't want them to walk away from the PR as a result. I think what would be a better situation would be for 1) the issue to be noticed during a precommit CI run, 2) the PR author to try to address the issues themselves, 3) the impacted project to be willing to step in and help out if needed (discuss solutions, offer to push changes to the PR to solve the issue, etc). Basically, everyone working together to try to get the patch landed with as little disruption or overhead as possible.

(3) comes with a fair amount of friction though because it's not always obvious *who* to reach out to, and (particularly for new contributors) it can be awkward to reach out to strangers to impose on them. So I wonder if it would be possible (or make sense) to have a precommit CI failure to automatically add the relevant PR subscribers/reviewers group to the PR. e.g., if there's a PR with a `clang` label and it breaks an lldb test, then lldb reviewers get automatically added to the PR, but if there's a PR with a `clang` and an `lldb` label and it breaks an lldb test, no new reviewers are added. This should hopefully reduce the friction for (3) because folks no longer need to figure out who to talk to. It should also hopefully reduce the amount of time spent churning on the review because the newly added reviewers can hopefully give some advice on how to resolve the issue (or just push a change to resolve it for the PR author).

If we had something like that, then I think we could manually exclude the handful of sometimes flaky tests but otherwise add lldb precommit CI to Clang's pipeline.

**Comment by mizvekov** [comment]
I would normally have no issue making the changes needed to other projects, and I mostly try to do that.
But lldb sets itself apart in how hard it is to get a project setup going, and how impactful it is.

It's a bit concerning when the changes needed are quite trivial, for example just changing diagnostics expectations, but you have no other way to test that.

I think the lldb pre-commit CI would at least help here, for trivial changes, we can keep trying to fix and verifying it by pushing to the PR, no changes needed on developer setup side.

**Comment by JDevlieghere** [comment]
> But lldb sets itself apart in how hard it is to get a project setup going, and how impactful it is.

Can you elaborate on what specifically you find hard? Building should be fairly straightforward, besides Python and SWIG, which are required to run the test, most of our dependencies are optional. Unlike the compiler, the debugger is (1) tied to the system it's running on, and (2) an interactive tool, so maybe that's what you're referring to? Either way, if there are concrete ways we can make LLDB easier to work on, I'd love to hear your suggestions!

**Comment by mizvekov** [comment]
> Can you elaborate on what specifically you find hard? Building should be fairly straightforward, besides Python and SWIG, which are required to run the tests, most of our dependencies are optional. Unlike the compiler, the debugger is (1) tied to the system it's running on, and (2) an interactive tool, so maybe that's what you're referring to? Either way, if there are concrete ways we can make LLDB easier to work on, I'd love to hear your suggestions!

This was about a year ago, and I remember @erichkeane had similar issues, but the main points:
1) Windows support
2) Build system does not support Multi-config generators. You can grep for https://cmake.org/cmake/help/latest/variable/CMAKE_CFG_INTDIR.html as an example.
3) Not all tests pass and it's quite unclear when that is not your fault.

AFAIK, on windows I would run into this issue where running the LLDB test suite would cause hundreds of frozen terminal windows to spawn, and that was quite difficult to recover from without rebooting.
This might be related to 2) above, before I figured it out.

I eventually thought that was too dangerous to run on windows and managed to get it working on linux virtual machine.

**Comment by medismailben** [comment]
> if there's a PR with a clang label and it breaks an lldb test, then lldb reviewers get automatically added to the PR

I think it would be reasonable to have the CI add the last (or last few) git author(s) for the failing test, to help investigate the issue on the lldb side.

**Comment by JDevlieghere** [comment]
> 1. Windows support

Windows support is definitely lacking which probably means a bunch of tests don't run there. We do have [one buildbot](https://lab.llvm.org/buildbot/#/builders/219) that runs on Windows so I have a hard time believing it's totally broken. The last runs are consistently green and based on the number of tests it seems to still cover a remarkable subset of the test suite. 

> 2. Build system does not support Multi-config generators. You can grep for https://cmake.org/cmake/help/latest/variable/CMAKE_CFG_INTDIR.html as an example.

This simply isn't true. Xcode is a multi-config build system and we have [a bot](https://ci.swift.org/view/all/job/llvm.org/view/LLDB/job/lldb-cmake-standalone/) that ensures that doesn't break. Maybe there's something different about MSVC? 

> 3. Not all tests pass and it's quite unclear when that is not your fault.

I really sympathize with this one. Most of that's boils down to how much we have to rely on the OS and other tools (the compiler, the linker, etc) for our integration tests. Having more tests written as unit tests or regression tests using `lldb-test` will help with that. I think we (the lldb community) can be more vigilant on that front in pushing for those tests. 👍

**Comment by slydiman** [comment]
We have configured few buildbots for cross lldb tests with Windows x86_64, Linux x86_64 hosts and Linux Aarch64 targets. Our main setup is Windows host and currently it is 100% green. We still have a number of local patches. The biggest changes are related to Makefile.rules to build API tests with different tools and hosts and linking the correct libraries on Linux host. We also adapted Shell tests to run on a remote target. We will prepare PRs to mainline soon. We have fugured out that currently lldb does not support CXX ABI 2. Few tests are marked as XFAIL with unclear reasons and strange conditions, so we got few XPASS tests. Most failed API tests on Windows hosts were related to path separators. Currently SBFileSpec (FileSpec wrapper for python) does not have a constructor to pass tripple or style, so it is impossible to create a POSIX path in API tests on the Windows host. But FileSpec stores a normalized path and the path style is completelly ignored inside lldb code.

**Comment by Endilll** [comment]
@AaronBallman @JDevlieghere @Michael137 @adrian-prantl @DavidSpickett @mizvekov I updated the PR to test LLDB on Clang changes. I hope we can start gathering approvals to get this PR moving.

File: clang/examples/PrintFunctionNames/PrintFunctionNames.cpp
Comment (JDevlieghere) [inline clang/examples/PrintFunctionNames/PrintFunctionNames.cpp:75]: Unrelated whitespace change?

**Review by JDevlieghere — APPROVED** [review]

File: clang/examples/PrintFunctionNames/PrintFunctionNames.cpp
Comment (Endilll) [reply clang/examples/PrintFunctionNames/PrintFunctionNames.cpp:75]: I need to make some changes in Clang for testing purposes. This will be removed before merging.

**Comment by AaronBallman** [comment]
I have no objections in theory, but in practice I remain concerned about the flakiness of lldb's tests for the impacts on Clang reviewers.

There are timeout failures, but they might be limited to Windows and thus not a problem with your PR (https://lab.llvm.org/buildbot/#/builders/219/builds/11819 and
https://lab.llvm.org/buildbot/#/builders/219/builds/11810 as examples).

But then there are failures like:
https://lab.llvm.org/buildbot/#/builders/17/builds/53939 -- this same stack trace seems to pop up in other failures and looks to be unrelated to changes in the change set, such as https://lab.llvm.org/buildbot/#/builders/68/builds/75367

or

https://lab.llvm.org/buildbot/#/builders/68/builds/75698 -- different assertion, but like above, seems to be spurious and unrelated to changes in the changeset, such as https://lab.llvm.org/buildbot/#/builders/68/builds/75666

Perhaps one way forward is for lldb tests to run on lldb changes for a few months before enabling them on Clang changes? Then the lldb folks can either report back that things are in pretty good shape from their experience, or we'll maybe have a better idea of how to proceed.

That said, if others aren't as concerned about test flakiness, I won't block progress on this PR.

**Comment by Endilll** [comment]
From my side as an author of the PR, I want to see Clang and LLDB code owners agree on the way forward, then I'll implement it if I can.

**Comment by philnik777** [comment]
FYI in the libc++ CI we've been running the LLDB tests for a few months now and I haven't seen a single unrelated failure.

**Comment by AaronBallman** [comment]
> FYI in the libc++ CI we've been running the LLDB tests for a few months now and I haven't seen a single unrelated failure.

Thanks, that's really useful information!

**Review by AaronBallman — APPROVED** [review]
Given the libc++ experience, let's move forward with this PR as-is. If there's an issue in practice, we'll disable the tests on changes to Clang at that point.

**Comment by Michael137** [comment]
> Given the libc++ experience, let's move forward with this PR as-is. If there's an issue in practice, we'll disable the tests on changes to Clang at that point.

Do note that libc++ CI only runs a subset of the LLDB tests. Only the ones that are impacted by libc++. We could take the same approach for clang if we find the whole test-suite to be too flakey i suppose

**Comment by mizvekov** [comment]
@Endilll This looks good, thanks!

How will this affect test capacity? Right now, the Linux bots are lagging, while the Windows bot is breezing through.

This is the opposite of the usual. Are we under-provisioned on Linux CI resources? How much worse will it get when we add this extra workload?

**Comment by Endilll** [comment]
> How will this affect test capacity?

Evidence is anecdotal, but we're looking at about 4 more minutes on Linux CI.

**Review by mizvekov — APPROVED** [review]

**Pull request merged**
